src/hooks/usePermit2Allowance.ts

```
@@ -0,0 +1,122 @@
+        if (state === ApprovalState.PENDING && isApprovalSyncing) {
+          return ApprovalState.SYNCING
+        } else if (state === ApprovalState.SYNCING && !isApprovalSyncing) {
+          return ApprovalState.SYNCED
+        }
+        return state
+      })
+    }
+  }, [isApprovalPending, isApprovalSyncing])
+
+  // Signature and PermitAllowance will expire, so they should be rechecked at an interval.
+  const [now, setNow] = useState(Date.now())
+  // Calculate now such that the signature will still be valid for the submitting block.
+  useInterval(() => setNow((Date.now() - STANDARD_L1_BLOCK_TIME) / 1000), STANDARD_L1_BLOCK_TIME, true)
+
+  const [signature, setSignature] = useState<PermitSignature>()
+  const isSigned = useMemo(() => {
+    if (!amount || !signature) return false
+    return signature.details.token === token?.address && signature.spender === spender && signature.sigDeadline >= now
+  }, [amount, now, signature, spender, token?.address])
+
+  const { permitAllowance, expiration: permitExpiration, nonce } = usePermitAllowance(token, account, spender)
+  const updatePermitAllowance = useUpdatePermitAllowance(token, spender, nonce, setSignature)
+  const isPermitted = useMemo(() => {
+    if (!amount || !permitAllowance || !permitExpiration) return false
+    return (permitAllowance.greaterThan(amount) || permitAllowance.equalTo(amount)) && permitExpiration >= now
+  }, [amount, now, permitAllowance, permitExpiration])
+
+  const callback = useCallback(async () => {
+    let info: ApprovalTransactionInfo | undefined
+    if (!(isApproved || isApprovalLoading)) {
+      info = await updateTokenAllowance()
+    }
+    if (!(isPermitted || isSigned)) {
+      await updatePermitAllowance()
+    }
+    return info
+  }, [isApprovalLoading, isApproved, isPermitted, isSigned, updatePermitAllowance, updateTokenAllowance])
+
+  return useMemo(() => {
+    if (token) {
+      if (!tokenAllowance || !permitAllowance) {
```

personally think its better to throw here than to keep the unknown state, as an invariant

---

unknown state can and will occur while the data is loading, and throwing would crash the application. this is a valid (but unknown) state, so it should not throw.

---

Renamed this to `AllowanceState.LOADING` to make this more clear (and not interpretable as an unhandled error state).